File: 2023/kotlin/src/day-02/Day02.kt
fun main() {

    val nonDigitRegex = "[^0-9]".toRegex()

    fun part1(input: List<String>): Int {

        val redLimit = 12
        val greenLimit = 13
        val blueLimit = 14

        var result = 0
        game@ for ((index, game) in input.withIndex()) {

            val grabs = game.split(":").last().split(";")
            for (grab in grabs) {

                val colors = grab.split(",")
                for (color in colors) {

                    val digit = color.replace(nonDigitRegex, "").toInt()

                    if (color.contains("red") && digit > redLimit) {
                        continue@game
                    } else if (color.contains("green") && digit > greenLimit) {
                        continue@game
                    } else if (color.contains("blue") && digit > blueLimit) {
                        continue@game
                    }
                }
            }

            result += (index + 1)
        }

        return result
    }

    fun part2(input: List<String>): Int {

        var result = 0
        game@ for (game in input) {

            var redMax = 0
            var greenMax = 0
            var blueMax = 0

            val grabs = game.split(":").last().split(";")
            for (grab in grabs) {

                val colors = grab.split(",")
                for (color in colors) {

                    val digit = color.replace(nonDigitRegex, "").toInt()

                    if (color.contains("red") && digit > redMax) {
                        redMax = digit
                    } else if (color.contains("green") && digit > greenMax) {
                        greenMax = digit
                    } else if (color.contains("blue") && digit > blueMax) {
                        blueMax = digit
                    }
                }
            }

            result += (redMax * greenMax * blueMax)
        }

        return result
    }

    val input = readInput("day-02/input")
    part1(input).println()
    part2(input).println()
}

File: 2023/go/day-02/day-02.go
package main

import (
	"fmt"
	utils "github.com/markusthielker/advent-of-code/2023/go"
	"regexp"
	"strconv"
	"strings"
)

func main() {

	nonDigitRegex := regexp.MustCompile(`[^0-9]`)

	part1 := func(input []string) (result int) {

		result, redLimit, greenLimit, blueLimit := 0, 12, 13, 14

	game:
		for index, game := range input {

			grabs := strings.Split(strings.Split(game, ":")[1], ";")
			for _, grab := range grabs {

				colors := strings.Split(grab, ",")
				for _, color := range colors {

					digit := nonDigitRegex.ReplaceAllString(color, "")
					digitTyped, _ := strconv.Atoi(digit)

					if strings.Contains(color, "red") && digitTyped > redLimit {
						continue game
					} else if strings.Contains(color, "green") && digitTyped > greenLimit {
						continue game
					} else if strings.Contains(color, "blue") && digitTyped > blueLimit {
						continue game
					}
				}
			}

			result += index + 1
		}

		return
	}

	part2 := func(input []string) (result int) {

		result = 0

		for _, game := range input {

			redMax, greenMax, blueMax := 0, 0, 0

			grabs := strings.Split(strings.Split(game, ":")[1], ";")
			for _, grab := range grabs {

				colors := strings.Split(grab, ",")
				for _, color := range colors {

					digit := nonDigitRegex.ReplaceAllString(color, "")
					digitTyped, _ := strconv.Atoi(digit)

					if strings.Contains(color, "red") && digitTyped > redMax {
						redMax = digitTyped
					} else if strings.Contains(color, "green") && digitTyped > greenMax {
						greenMax = digitTyped
					} else if strings.Contains(color, "blue") && digitTyped > blueMax {
						blueMax = digitTyped
					}
				}
			}

			result += redMax * greenMax * blueMax
		}

		return
	}

	input := utils.ReadInput("day-02/input")

	fmt.Println(part1(input))
	fmt.Println(part2(input))
}
